### DISCUSSION TITLE
Accessing Javascript library from web project

### DISCUSSION THREAD

**Discussion by user72356** [mapeditor/tiled#3181] [General]
I've been developing custom Tiled extensions for my project using the Javascript API (BTW great API, very powerful!). Eventually I will also need to perform similar operations as these extensions, but from a web app instead.

I'm not 100% sure how the Javascript API works, but if I had to guess it's probably a thin layer into the C++ libtiled? Meaning that Tiled itself needs to be running for this API to function? I'm wondering how I could read a map, perform operations using the API, and then save the map, all from a web app, i.e. I can't open Tiled from the front-end client.

There are libs out there for various languages and frameworks that can read and write (?) the TMX files, but I don't think they have functionality such as TileAt, NewLayer, etc.

Thanks for any pointers!

**Comment by bjorn** [comment]
> I'm not 100% sure how the Javascript API works, but if I had to guess it's probably a thin layer into the C++ libtiled? Meaning that Tiled itself needs to be running for this API to function?

Yeah, it's a thin layer into C++ Tiled, so it's not even out in its own library at the moment (something that may change, but it's unlikely going to help make this API available on the web). It sounds like you're going to need a JS library, but I'm not sure if there are any that can also write TMX files.

If you're going to write your own solution, note that it might be easier to go with JSON, which is more straight-forward to read and write in JavaScript. The JSON format in Tiled supports all features so it is a full replacement for the TMX / TSX formats. We could likely even set up a type definitions file for the JSON format, based on which you could generate JavaScript or TypeScript code using [quicktype](https://app.quicktype.io/) that can load and save the format.

I think about the only tricky part, if you're going to read & write these files, would be to convert the global tile IDs to tile references after loading, and back to global tile IDs before saving. There's a [page about the global tile IDs](https://doc.mapeditor.org/en/latest/reference/global-tile-ids.html) in the manual.

**Reply by user72356** [reply]
I'm not dead-set with JS since I'm writing my app in C# (Blazor), but can use JS as well. I know there is a C# lib, but that also is limited to reading maps.

Maybe if there was a command line interface to Tiled that would expose the C++ API? This would make working with maps universally accessible. For example, in a web app, the client could issue a request into the server, and the server would in turn use the command line interface to interact with the map. I haven't looked at libtiled but this could be a simple 1-to-1, thin layer from the command line to libtiled. We could even use JSON as the input/output mechanism to represent complex data. This would be really cool as a read/write/interactive mechanism straight into the core of Tiled, and no need for unofficial 3rd party libs. I can envision this as an interactive command line similar to telnet.

I'm stretched extremely thin but maybe I can drop this as a feature request!! (Thank you for your attention ;-) )
